common/config/src/raw.rs

```diff
@@ -70,6 +70,19 @@ where
     d.deserialize_option(OptionVisitor)
 }
 
+fn params_deser<'de, D>(data: D) -> Result<Option<Params>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let mut params_builder: ParamsBuilder = Deserialize::deserialize(data)?;
+    Ok(params_builder.build().ok().or_else(|| {
+        params_builder
+            .hostname(get_hostname().unwrap_or_default())
+            .build()
+            .ok()
```

Comment:
Defaulting only if it can't parse the initial value from file?

Reply:
Exactly.

It derualises the params field, if there is one, into a ParamsBuilder and then uses the Builder's built.in validation logic to check if the required fields are populated (ie hostname) if hostname is not provided it sets hostname and then builds a params object from the now fully populated builder.

Any otger implementation I could think of either required hostname in order to populate params, which is bad as you may well want to populate params.tags and default the hostname (which we already do for env var parsing).